coral-schema/src/main/java/com/linkedin/coral/schema/avro/RelToAvroSchemaConverter.java

```diff
@@ -459,31 +460,41 @@ public RexNode visitRangeRef(RexRangeRef rexRangeRef) {
     @Override
     public RexNode visitFieldAccess(RexFieldAccess rexFieldAccess) {
       RexNode referenceExpr = rexFieldAccess.getReferenceExpr();
-      Deque<String> innerRecordNames = new LinkedList<>();
-      while (!(referenceExpr instanceof RexInputRef)) {
-        if (referenceExpr instanceof RexCall
-            && ((RexCall) referenceExpr).getOperator().getName().equalsIgnoreCase("ITEM")) {
-          // While selecting `int_field` from `array_col:array<struct<int_field:int>>` using `array_col[x].int_field`,
-          // `rexFieldAccess` is like `ITEM($1, 1).int_field`, we need to set `referenceExpr` to be the first operand (`$1`) of `ITEM` function
-          referenceExpr = ((RexCall) referenceExpr).getOperands().get(0);
-        } else if (referenceExpr instanceof RexFieldAccess) {
-          // While selecting `int_field` from `struct_col:struct<inner_struct_col:struct<int_field:int>>` using `struct_col.inner_struct_col.int_field`,
-          // `rexFieldAccess` is like `$3.inner_struct_col.int_field`, we need to set `referenceExpr` to be the expr (`$3`) of itself.
-          // Besides, we need to store the field name (`inner_struct_col`) in `fieldNames` so that we can retrieve the correct inner struct from `topSchema` afterwards
-          innerRecordNames.push(((RexFieldAccess) referenceExpr).getField().getName());
-          referenceExpr = ((RexFieldAccess) referenceExpr).getReferenceExpr();
-        } else {
-          return super.visitFieldAccess(rexFieldAccess);
-        }
-      }
       String oldFieldName = rexFieldAccess.getField().getName();
       String suggestNewFieldName = suggestedFieldNames.poll();
       String newFieldName = SchemaUtilities.getFieldName(oldFieldName, suggestNewFieldName);
```

Comment:
Did some investigations, the regression is because the `suggestedFieldNames` is empty so that `newFieldName` is null and causes NPE:
```
java.lang.NullPointerException
	at com.linkedin.coral.com.google.common.base.Preconditions.checkNotNull(Preconditions.java:212)
	at com.linkedin.coral.schema.avro.SchemaUtilities.getFieldName(SchemaUtilities.java:264)
	at com.linkedin.coral.schema.avro.RelToAvroSchemaConverter$SchemaRexShuttle.visitFieldAccess(RelToAvroSchemaConverter.java:465)
	at com.linkedin.coral.schema.avro.RelToAvroSchemaConverter$SchemaRexShuttle.visitFieldAccess(RelToAvroSchemaConverter.java:370)
	at org.apache.calcite.rex.RexFieldAccess.accept(RexFieldAccess.java:81)
	at org.apache.calcite.rex.RexShuttle.apply(RexShuttle.java:277)
	at org.apache.calcite.rex.RexShuttle.mutate(RexShuttle.java:239)
	at org.apache.calcite.rex.RexShuttle.apply(RexShuttle.java:257)
	at org.apache.calcite.rel.core.Project.accept(Project.java:144)
	at com.linkedin.coral.schema.avro.RelToAvroSchemaConverter$SchemaRelShuttle.visit(RelToAvroSchemaConverter.java:209)
	at org.apache.calcite.rel.logical.LogicalProject.accept(LogicalProject.java:121)
	at org.apache.calcite.rel.RelShuttleImpl.visitChild(RelShuttleImpl.java:55)
	at org.apache.calcite.rel.RelShuttleImpl.visitChildren(RelShuttleImpl.java:69)
	at org.apache.calcite.rel.RelShuttleImpl.visit(RelShuttleImpl.java:107)
	at com.linkedin.coral.schema.avro.RelToAvroSchemaConverter$SchemaRelShuttle.visit(RelToAvroSchemaConverter.java:244)
	at org.apache.calcite.rel.logical.LogicalCorrelate.accept(LogicalCorrelate.java:109)
	at org.apache.calcite.rel.RelShuttleImpl.visitChild(RelShuttleImpl.java:55)
	at org.apache.calcite.rel.RelShuttleImpl.visit(RelShuttleImpl.java:99)
	at com.linkedin.coral.schema.avro.RelToAvroSchemaConverter$SchemaRelShuttle.visit(RelToAvroSchemaConverter.java:199)
	at org.apache.calcite.rel.logical.LogicalProject.accept(LogicalProject.java:121)
	at com.linkedin.coral.schema.avro.RelToAvroSchemaConverter.convert(RelToAvroSchemaConverter.java:126)
	at com.linkedin.coral.schema.avro.ViewToAvroSchemaConverter.inferAvroSchema(ViewToAvroSchemaConverter.java:152)
	at com.linkedin.coral.schema.avro.ViewToAvroSchemaConverter.toAvroSchema(ViewToAvroSchemaConverter.java:90)
	at com.linkedin.coral.tools.AvroSchemaValidator.toAvroSchema(AvroSchemaValidator.java:43)
	at com.linkedin.coral.tools.AvroSchemaValidator.convertAndValidate(AvroSchemaValidator.java:33)
	at com.linkedin.coral.tools.AllViewTranslation.convertView(AllViewTranslation.java:209)
	at com.linkedin.coral.tools.AllViewTranslation.translateAllViews(AllViewTranslation.java:152)
	at com.linkedin.coral.tools.AllViewTranslation.main(AllViewTranslation.java:112)
```
I am not sure if we should guarantee that `suggestedFieldNames` is never empty. If so, we may need to do more investigations; if not, I think one option is to modify the `getFieldName` to allow the null `suggestedNewName`

Reply:
We should guarantee that `suggestedFieldNames` is never empty. The exception is due to `suggestedFieldNames ` is polled at wrong place. This is fixed in the latest commit.